### DISCUSSION TITLE
Additional Scopes

### DISCUSSION THREAD

**Discussion by dickhardt** [hellocoop/hello.dev#4] [Ideas]
We are exploring adding the following scopes:

|Scope|Description|
|---|---|
|`github`|The user's GitHub profile url.|
|`linkedin`|The user's LinkedIn profile url.|
|`ethereum` |The user's ethereum address.|

Would these be useful? Other scopes you would like to see?

**Comment by dickhardt** [comment]
We would prompt the user to link their GitHub, LinkedIn accounts, or their Ethereum Wallet.

Q: Would we require the user to complete the process before sending them back to your app? Or if they cancel, we would not include the requested scope in the `scope` claim in the ID Token? There is no simple way of indicating if a scope is optional vs required.

**Reply by ltd** [reply]
I'm +1 on this, and more in the future.  It's huge value for the developer.   Why not have dual flavors,  required_github & maybe_github, for indicating if the scope is mandatory.

**Reply by dickhardt** [reply]
It seems you wanted a mandatory handle that could be either - ie one of GitHub or gitlab 

To be spec compliant we would move to Json to express optional claims -- but I consider that a big step up for developers to understand

gitlab-optional was the scope format I was considering - but does not address the one or other use case

A simpler approach would be the app would offer the user to pick if they wanted to link github or gitlab - and then the scope is required at Hellō - ie user cancels or links a github account

**Comment by dickhardt** [comment]
We now support github, gitlab, twitter, discord where we return a JSON object 
```json
{
  "id": "<unique_user_id_from_service>",
  "handle":"<user_selected_handle_that_may_change>"  
}
```

**Comment by aslakg** [comment]
I'm trying to use this with rabbitmq,  and it requires custom scopes: https://github.com/rabbitmq/rabbitmq-server/tree/main/deps/rabbitmq_auth_backend_oauth2

**Comment by dickhardt** [comment]
@aslakg you are wanting an access token (authorization) -- Hellō provides ID Tokens which represent who the user is, not what the user is allowed to do.

**Reply by aslakg** [reply]
Yes, unfortunately rabbitmq does both with the same mechanism. But I found a work around by aliasing the "openid" scope to their requirements. So feel free to delete my comments.

**Reply by dickhardt** [reply]
Glad you found a work around. Will leave your comment here for now as it may be useful to someone else.

**Comment by jimkleiber** [comment]
For years I've been wanting a way to verify the humanity of users on my platform. As AI takes off, I feel even more desire to be able to say "this account is matched to this verified human."

There are ways to verify identity through Stripe Identity and others, where they check one's passport or ID, but often these solutions provide all of the ID information to the platform.

I don't want all of that information, I just want to know if they're a verified human or not.

Is there any way to provide a scope for verified human? Or even verified name that matches their government ID?

I think some services would be interested in other information from that ID verification process, like verified 18+ age, or verified passport number for airline websites, but those aren't so important to me now.

If Hellō is to become an identity wallet, I think incorporating verified physical identities might help it stand out from the other SSO providers. I would trust Hellō with my passport info, not Google and Facebook and LinkedIn and less secure websites who ask.

What do you think?

**Reply by dickhardt** [reply]
Hey Jim, we envision providing a claim that the account is a real person. It is not as simple as it may seem. A couple of examples: a human may have enrolled the account, and gave it to a bot -- or the account was compromised and is now controlled by a bot.

Would love to learn what you are wanting to prevent, or what you want to know. Would you elaborate?

**Reply by jimkleiber** [reply]
Heyyy, mostly just trying to help people on a forum or new social media know that they're interacting with an account that has been verified to be owned by a human. It doesn't necessarily prove that the one using the account is a human, as you said, the human could outsource the typing to a bot or LLM, but at least there will be somewhat of an accountability trail.

**Reply by jimkleiber** [reply]
For example, on Instagram, I know a verified profile is owned by a specific person. I don't know if they're the one posting the photos there, but I know it's at least under their jurisdiction.

**Reply by dickhardt** [reply]
Most easy things will be a proxy or building off of a different service that does a decent job of bot detection such as Google.

One of the benefits we believe we will get from B2B support is businesses tend to do a good job of ensuring that their employees are human!

**Reply by jimkleiber** [reply]
Yes, i agree, as facebook grew strongly off colleges verifying student identities through the college email address.

Just was wondering, from the side as an individual user of Hellō, not from the side of a dev, how nice it'd be to have a digital identity wallet where I could have my actual government issued verified/stored and i could choose which info i give to which sites. Maybe Apple will get there with putting ID cards into its wallet but this is another angle i see

**Reply by jimkleiber** [reply]
Here's an example I found of Verify with Apple Wallet:

https://medium.com/@drewsmith_6943/apple-wallet-id-is-the-solution-to-age-verification-online-26bef8219832

**Reply by dickhardt** [reply]
Would you elaborate on the article you linked to?

Few people have an mDL on their phone, it is only available for iOS apps, and Apple heavily restricts which apps can access it -- so usage is nominal.

**Reply by jimkleiber** [reply]
For me it's not so much about Apple's current implementation of it, but rather a growing desire (by some) to have verified claims of identity on the internet. KYC has been there for a while in the banking sector, but there's growing desire in the UK and Europe for age verification and other verified claims.

I've seen other services that provide this, whether ID.me, Yoti, oneID.co.uk, etc., even providing claims like "over 18?" as a boolean.

I guess I just mentioned it because I remember you talking of verified claims before and I'm just curious/excited to see how the verified claims will work through Hellō. OneID, for example, charges 54p per identity verification and 16p per age verification, and as a developer, I could currently connect my app to OneID directly and pay that, or ideally, I could just go through Hellō and not have to deal with tons of OIDC providers or APIs.

**Reply by dickhardt** [reply]
Got it -- verified identity is on roadmap -- and we expect our pricing to be disruptive!

**Reply by jimkleiber** [reply]
I am very excited to hear that 🙂🙂🙂

And again, thank you for Hellō, I love it.

**Reply by jimkleiber** [reply]
Another (much simpler) request. Is there a scope for a short bio/description?

Almost every social media site uses short profile bios/descriptions and I think it'd be really nice to be able to manage that from my Hellō wallet.

Also I see Twitter and Github and would make a request for Instagram and LinkedIn. And if we're asking for a lot, then Bluesky and Threads as well.

The reason for the social media ones is to kinda do Keybase did back in the day: better prove one's identity by showing more proof of the accounts they control.

Lastly, in this vein, to maybe show if they manage a specific domain name. Not sure if this could be a scope tho.

**Comment by dickhardt** [comment]
We had bios when we were doing Mastodon -- the length of a bio and what goes in a given site sometimes are similar -- but people often have different ones for different sites

`linkedin` and `instagram` don't allow us to get a users handle unfortunately -- threads is in the same boat -- we might be able to do BlueSky -- but it social is not a priority at this time -- Twitter has yet to be asked for

**Reply by jimkleiber** [reply]
Hmm. I think with Clubhouse back in the day, when someone logged into Instagram thru Clubhouse, they were able to show the verified IG handle, but maybe I'm mistaken.

Really just trying to say "this Hellō account owns this IG account" but maybe that's not possible right now.

Yes bios can be different lengths and have different requirements, which might be annoying. And yes people can prefer different ones for different sites. But a huge benefit I see in Hellō is being able to write a few different ones and choose which one I wanna use and where I wanna use it. Just how nicknames and photos work right now.

What about preferred username?

Again, different sites have different requirements for this, too, but I'd hope some would standardize a bit over time, especially if Hellō takes off.

**Comment by dickhardt** [comment]
not sure why we are not exposing `preferred_username` as a scope -- we are gathering it
